craft_parts/executor/step_handler.py

```diff
@@ -0,0 +1,361 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright 2016-2021 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Handle the execution of built-in or user specified step commands."""
+
+import fileinput
+import json
+import os
+import re
+import subprocess
+import sys
+import tempfile
+import textwrap
+import time
+from collections import namedtuple
+from pathlib import Path
+from typing import List, Optional, Set, Union
+
+from craft_parts import errors
+from craft_parts.executor import collisions
+from craft_parts.infos import StepInfo
+from craft_parts.parts import Part
+from craft_parts.plugins import Plugin
+from craft_parts.sources import SourceHandler
+from craft_parts.steps import Step
+from craft_parts.utils import file_utils
+
+from . import environment, filesets
+from .filesets import Fileset
+
+FilesAndDirs = namedtuple("FilesAndDirs", ["files", "dirs"])
```

Comment:
```suggestion
@dataclasses.dataclass
class FilesAndDirs:
  files: Set[str]
  dirs: Set[str]
```

Reply:
In this case we're simply organizing the tuple returned by handler functions, and adding the overhead of a full dataclass is probably not worth it.

Reply:
> In this case we're simply organizing the tuple returned by handler functions, and adding the overhead of a full dataclass is probably not worth it.

The advantage to the dataclass over namedtuple is type information.  Which could be helpful in development.  But not really critical.